fairscale/experimental/wgit/weigit_api.py

```diff
@@ -0,0 +1,172 @@
+        # and notes all the internal files.
+        # else, if repo already exists: create a pygit object from the .wgit/.git.
+        self.wgit_parent = parent_dir
+        self._repo_path: Union[None, Path] = None
+        self._wgit_dir = Path(".wgit")
+        self._metadata_file = Path(".wgit/checkpoint.pt")
+        self._sha1_ref = Path(".wgit/sha1_refs.json")
+        self._wgit_git_path = Path(".wgit/.git")
+        self._sha1_store = Path(".wgit/sha1_store")
+
+        if not self._exists() and init:
+            # No weigit repo exists and is being initialized with init=True
+            # Make .wgit directory, create sha1_refs and metadata file
+            create_dir(dir_path=self._wgit_dir, exception_msg="An exception occured: WeiGit already Initialized")
+            create_file(file_path=self._metadata_file)
+            create_file(file_path=self._sha1_ref)
+
+            # Make the .wgit a git repo
+            try:
+                pygit2.init_repository(str(self._wgit_git_path), False)
+                self.pygit = PyGit(self.wgit_parent.joinpath(self._wgit_dir))
+                create_file(self._wgit_dir.joinpath(".gitignore"))
+                write_to_file(
+                    self._wgit_dir.joinpath(".gitignore"), msg=f"{self._sha1_ref.name}\n{self._sha1_store.name}"
+                )
+            except pygit2.GitError as error:
+                sys.stderr.write(f"An exception occurred while initializing .wgit/.git: {repr(error)}\n")
+                sys.exit(ExitCode.ERROR)
+
+            # Initializing sha1_store only after wgit has been initialized!
+            SHA1_store()
+        elif init:
+            # if weigit repo already exists and init is being called, wrap the existing .wgit/.git repo with PyGit
+            self.pygit = PyGit(self.wgit_parent.joinpath(self._wgit_dir))
+        elif not self._exists() and not init:
+            # weigit doesn't exist and is not trying to be initialized (triggers during non-init commands)
+            sys.stderr.write("fatal: not a wgit repository!\n")
+        else:
+            # weigit exists and non-init commands are triggered
+            self.pygit = PyGit(self.wgit_parent.joinpath(self._wgit_dir))
+
+    def add(self, file_path: str) -> None:
+        if self._exists():
+            # file_name = os.path.basename(file_path)
+            sha1_hash = get_sha1_hash(file_path)
+
+            # use the sha1_has to create a directory with first2 sha naming convention
+            try:
+                repo_fdir = os.path.join(self.path, "sha1_store", sha1_hash[:2]) + "/"
+                os.makedirs(repo_fdir)
+            except FileExistsError as error:
+                sys.stderr.write(f"An exception occured: {repr(error)}\n")
+                sys.exit(ExitCode.FILE_EXISTS_ERROR)
+            try:
+                # First transfer the file to the internal sha1_store
+                repo_fpath = os.path.join(repo_fdir, sha1_hash[2:])
```

Comment:
this should probably be part of sha1_store.add() function.

Reply:
Ah I see, this makes is more reasonable. Need to update the design doc a bit as well to reflect this.